File: src/main/java/ua/notion/infrastructure/persistence/contract/PlantInstanceRepository.java
package ua.notion.infrastructure.persistence.contract;

import java.util.List;
import ua.notion.domain.entity.PlantInstance;
import ua.notion.infrastructure.persistence.Repository;

public interface PlantInstanceRepository extends Repository<PlantInstance, Integer> {

  List<PlantInstance> findByGardenId(Integer gardenId);

  List<PlantInstance> findByPlantId(Integer plantId);

  List<PlantInstance> findByGardenIdAndCellPosition(Integer gardenId, Integer cellX, Integer cellY);

  long countByGardenId(Integer gardenId);
}

File: src/server/src/repositories/contract/plant_instance_repository.rs
use axum::async_trait;

use crate::{errors::error::DbResult, models::PlantInstance, repositories::repository::Repository};

#[async_trait]
pub trait PlantInstanceRepository: Repository<PlantInstance, i32> {
    async fn find_by_garden_id(&self, garden_id: i32) -> DbResult<Vec<PlantInstance>>;
    async fn find_by_plant_id(&self, plant_id: i32) -> DbResult<Vec<PlantInstance>>;
    async fn find_by_position(
        &self,
        garden_id: i32,
        cell_x: i32,
        cell_y: i32,
    ) -> DbResult<Option<PlantInstance>>;
    async fn exists_at_position(&self, garden_id: i32, cell_x: i32, cell_y: i32) -> DbResult<bool>;
    async fn is_watered(&self, id: i32) -> DbResult<Option<bool>>;
    async fn is_fertilized(&self, id: i32) -> DbResult<Option<bool>>;
    async fn find_growth_stage_by_id(&self, id: i32) -> DbResult<Option<i32>>;
}
